lens/fetcher.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/fetcher"
+	"github.com/sourcenetwork/defradb/planner/mapper"
+)
+
+// todo: The code in here can be significantly simplified with:
+// https://github.com/sourcenetwork/defradb/issues/1589
+
+type lensedFetcher struct {
+	source   fetcher.Fetcher
+	registry client.LensRegistry
+	lens     Lens
+
+	txn datastore.Txn
+
+	col *client.CollectionDescription
+	// Cache the fieldDescriptions mapped by name to allow for cheaper access within the fetcher loop
+	fieldDescriptionsByName map[string]client.FieldDescription
+
+	targetVersionID string
+}
+
+var _ fetcher.Fetcher = (*lensedFetcher)(nil)
+
+// NewFetcher returns a new fetcher that will migrate any documents from the given
+// source Fetcher as they are are yielded.
+func NewFetcher(source fetcher.Fetcher, registry client.LensRegistry) fetcher.Fetcher {
+	return &lensedFetcher{
+		source:   source,
+		registry: registry,
+	}
+}
+
+func (f *lensedFetcher) Init(
+	col *client.CollectionDescription,
+	fields []client.FieldDescription,
+	filter *mapper.Filter,
+	docmapper *core.DocumentMapping,
+	reverse bool,
+	showDeleted bool,
+) error {
+	f.col = col
+
+	f.fieldDescriptionsByName = make(map[string]client.FieldDescription, len(col.Schema.Fields))
+	// Add cache the field descriptions in reverse, allowing smaller-index fields to overwrite any later
+	// ones.  This should never really happen here, but it ensures the result is consistent with col.GetField
+	// which returns the first one it finds with a matching name.
```

Comment:
Because if a single `GetField` call always returns the first matching (`a` in later example), an iterator needs to ensure that the first item is preserved instead of being overwritten by items later in the slice (`a` is not overwritten by `b`). Iterating in reverse ensures that `b` is overwritten by `a`.